traffic_ops/app/lib/API/CachegroupFallback.pm

```diff
@@ -0,0 +1,287 @@
+			&log( $self, "Backup configuration DELETED", "APICHANGE");
+			return $self->success_message("Backup configuration DELETED");
+		} else {
+			return $self->alert( "Backup configuration DELETED." );
+		}
+	} else {
+		&log( $self, "No backup Cachegroups found");
+		return $self->not_found();
+	}
+}
+
+sub show {
+	my $self = shift;
+	my $cache_id = $self->param("cacheGroupId");
+	my $fallback_id = $self->param("fallbackId");
+	my $id = $cache_id ? $cache_id : $fallback_id;
+
+	#only integers
+	if ( $id !~ /^\d+?$/ ) {
+		&log( $self, "No such Cachegroup id $id");
+		return $self->success([]);
+	}
+
+	my $cachegroup = $self->db->resultset('Cachegroup')->search( { id => $id } )->single();
+	if ( !defined($cachegroup) ) {
+		&log( $self, "No such Cachegroup $id");
+		return $self->success([]);
+	}
+
+	if ( ($cachegroup->type->name ne "EDGE_LOC") ) {
+		&log( $self, "cachegroup should be type EDGE_LOC.");
+		return $self->success([]);
+	}
+
+	my $rs_backups = undef;
+
+	if ( defined ($cache_id) && defined ($fallback_id)) {
+		$rs_backups = $self->db->resultset('CachegroupFallback')->search({ primary_cg => $cache_id, backup_cg => $fallback_id}, {order_by => 'set_order'});
+	} elsif ( defined ($cache_id) ) {
+		$rs_backups = $self->db->resultset('CachegroupFallback')->search({ primary_cg => $cache_id}, {order_by => 'set_order'});
+	} elsif ( defined ($fallback_id) ) {
+		$rs_backups = $self->db->resultset('CachegroupFallback')->search({ backup_cg => $fallback_id}, {order_by => 'set_order'});
+	}
+
+	if ( defined ($rs_backups) && ($rs_backups->count > 0) ) {
+		my $response;
+		my $backup_cnt = 0;
+		while ( my $row = $rs_backups->next ) {
+			$response->[$backup_cnt]{"cacheGroupId"} = $row->primary_cg->id;
+			$response->[$backup_cnt]{"cacheGroupName"} = $row->primary_cg->name;
+			$response->[$backup_cnt]{"fallbackName"} = $row->backup_cg->name;
+			$response->[$backup_cnt]{"fallbackId"} = $row->backup_cg->id;
+			$response->[$backup_cnt]{"fallbackOrder"} = $row->set_order;
+			$backup_cnt++;
+		}
+		return $self->success( $response );
+	} else {
+		&log( $self, "No backup Cachegroups");
+		return $self->success([]);
+	}
+}
+
+sub create {
+	my $self = shift;
+	my $cache_id = $self->param('cacheGroupId');
+	my $params = $self->req->json;
+
+	if ( !&is_oper($self) ) {
+		return $self->forbidden();
+	}
+
+	if ( !defined($params) ) {
+		return $self->alert("parameters must be in JSON format,  please check!");
+	}
+
+	if ( !defined($cache_id)) {
+		my @param_array = @{$params};
+		$cache_id = $param_array[0]{cacheGroupId};
+	}
+
+	my ( $is_valid, $result ) = $self->is_valid_cachegroup_fallback($params, $cache_id);
+
+	if ( !$is_valid ) {
+		return $self->alert($result);
+	}
+
+	foreach my $config (@{ $params }) {
+		my $rs_backup = $self->db->resultset('Cachegroup')->search( { id => $config->{fallbackId} } )->single();
+		if ( !defined($rs_backup) ) {
+			&log( $self, "ERROR Backup config: No such Cache Group $config->{fallbackId}");
```

Comment:
Ok. I have a doubt on this comment. Say i get a list of 10 fallbacks, where the 10th one is invalid. when i reach the 10th, i have already inserted the other 9. That is the reason i coded it this way so that only the valid elements will be inserted. Also , on a successful POST / PUT, the user can always do a GET to verify.

Reply:
If the client receives a 200, they should expect that the entire request was successful and shouldn't have to verify that all backups were actually created. Thinking about how we'd do this in the golang API, the entire thing would likely be wrapped in a transaction so that it creates either none or all of the requested backups. I'm ok with leaving this in the Perl as-is but definitely think it should be entirely wrapped in a transaction in the golang API.